partitioner/src/main/java/io/confluent/connect/storage/partitioner/TimeBasedPartitioner.java

```diff
@@ -103,8 +139,102 @@ public String encodePartition(SinkRecord sinkRecord) {
           (Class<? extends SchemaGenerator<T>>) config.get(PartitionerConfig.SCHEMA_GENERATOR_CLASS_CONFIG);
       return generatorClass.getConstructor(Map.class).newInstance(config);
     } catch (ClassCastException | IllegalAccessException | InstantiationException | InvocationTargetException
-        | NoSuchMethodException e) {
+                 | NoSuchMethodException e) {
       throw new ConfigException("Invalid generator class: " + generatorClass);
     }
   }
+
+  public TimestampExtractor newTimestampExtractor(String extractorClassName) {
+    try {
+      switch (extractorClassName) {
+        case "Wallclock":
+        case "Record":
+        case "RecordField":
+          extractorClassName = "io.confluent.connect.storage.partitioner.TimeBasedPartitioner$" + extractorClassName
+                                   + "TimestampExtractor";
+          break;
+        default:
+      }
+      Class<?> klass = Class.forName(extractorClassName);
+      if (!TimestampExtractor.class.isAssignableFrom(klass)) {
+        throw new ConnectException("Class " + extractorClassName + " does not implement TimestampExtractor");
+      }
+      return (TimestampExtractor) klass.newInstance();
+    } catch (ClassNotFoundException | ClassCastException | IllegalAccessException | InstantiationException e) {
+      throw new ConfigException("Invalid timestamp extractor: " + extractorClassName, e);
+    }
+  }
+
+  public static class WallclockTimestampExtractor implements TimestampExtractor {
+    @Override
+    public void configure(Map<String, Object> config) {}
+
+    @Override
+    public Long extract(ConnectRecord<?> record) {
+      return Time.SYSTEM.milliseconds();
+    }
+  }
+
+  public static class RecordTimestampExtractor implements TimestampExtractor {
+    @Override
+    public void configure(Map<String, Object> config) {}
+
+    @Override
+    public Long extract(ConnectRecord<?> record) {
+      return record.timestamp();
+    }
+  }
+
+  public static class RecordFieldTimestampExtractor implements TimestampExtractor {
+    private String fieldName;
+    private DateTimeFormatter dateTime;
+
+    @Override
+    public void configure(Map<String, Object> config) {
+      fieldName = (String) config.get(PartitionerConfig.TIMESTAMP_FIELD_NAME_CONFIG);
+      dateTime = ISODateTimeFormat.dateTime();
```

Comment:
I assumed the user-defined timestamp field won't need time zone conversion. But will revisit during tests.